### PR TITLE
fix(users): Add max wrong attempts for two-fa

### PR DESCRIPTION
**Pull request by Riddhiagrawal001**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [X] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Previously, there was no limit on the number of failed attempts for TOTP and recovery code. With this PR, a maximum of 4 wrong attempts is introduced for both TOTP and recovery code. After 4 failed attempts, a cooldown period of 5 minutes applies for TOTP, and 10 minutes for the recovery code.

- A new endpoint is introduced to check the no of attempts of the user and also if the user has done two-fa recently .

### Additional Changes

- [X] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Request : 
```
curl --location 'http://localhost:8080/user/2fa/v2' \
--header 'Authorization: Bearer JWT'
```

Response:
- If the user has not set twofa 
```
{
    "status": null
}
```
- If the user has twofa set and no attempts is remaining for both recovery code and totp 
```
{
    "status": {
        "totp": {
            "is_completed": false,
            "remaining_attempts": 0
        },
        "recovery_code": {
            "is_completed": false,
            "remaining_attempts": 0
        }
    }
}
```
- If the user has twofa set and no attempts is remaining for recovery code  
```
{
    "status": {
        "totp": {
            "is_completed": false,
            "remaining_attempts": 4
        },
        "recovery_code": {
            "is_completed": false,
            "remaining_attempts": 0
        }
    }
}
```
- If the user has twofa set and no attempts is remaining for totp
```
{
    "status": {
        "totp": {
            "is_completed": false,
            "remaining_attempts": 0
        },
        "recovery_code": {
            "is_completed": false,
            "remaining_attempts": 4
        }
    }
}
```

After 4 wrong recovery code attempts , `2fa/recovery_code/verify` will throw `UR_49` error
Request :
```
curl --location 'http://localhost:8080/user/2fa/recovery_code/verify' \
--header 'authorization: Bearer JWT' \
--data '{"recovery_code":"some_invalid_recovery_code"}'
```

Response :
```
{
    "error": {
        "type": "invalid_request",
        "message": "Maximum recovery code attempts per user reached",
        "code": "UR_49"
    }
}
```

After 4 wrong totp attempts , `2fa/recovery_code/verify` will throw `UR_48` error
Request :
```
curl 'http://localhost:8080/user/2fa/totp/verify' \
  -H 'authorization: Bearer JWT' \
  --data-raw '{"totp":"some_invalid_totp"}'
```
  
Response :
```
{
    "error": {
        "type": "invalid_request",
        "message": "Maximum totp attempts per user reached",
        "code": "UR_48"
    }
}
```
  

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/core/errors/user.rs [hunk 1]
```diff
@@ -73,6 +73,10 @@ pub enum UserErrors {
     #[error("TotpRequired")]
     TotpRequired,
     #[error("InvalidRecoveryCode")]
+    MaxTotpAttemptsReached,
+    #[error("MaxTotpAttemptsReached")]
+    MaxRecoveryCodeAttemptsReached,
+    #[error("MaxRecoveryCodeAttemptsReached")]
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:79]: It's better to place these at last and follow the order of error codes when defining errors.

And also change the error attribute message to match the one defined in `get_error_message` function.

File: crates/router/src/core/errors/user.rs [hunk 2]
```diff
@@ -269,6 +279,10 @@ impl UserErrors {
             Self::InvalidTotp => "Invalid TOTP",
             Self::TotpRequired => "TOTP required",
             Self::InvalidRecoveryCode => "Invalid Recovery Code",
+            Self::MaxTotpAttemptsReached => "Maximum totp attempts per user reached",
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:282]:
```suggestion
            Self::MaxTotpAttemptsReached => "Maximum TOTP attempts per user reached",
```

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 3]
```diff
@@ -50,6 +50,69 @@ pub async fn check_recovery_code_in_redis(state: &SessionState, user_id: &str) -
...
+        .set_key_with_expiry(
+            key.as_str(),
+            user_attempts_value + 1,
+            consts::user::REDIS_RECOVERY_CODE_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+
+pub async fn check_recovery_code_attempts_in_redis(
```
Comment (ThisIsMani) [inline crates/router/src/utils/user/two_factor_auth.rs:99]:
```suggestion
pub async fn get_recovery_code_attempts_from_redis(
```

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 4]
```diff
@@ -50,6 +50,69 @@ pub async fn check_recovery_code_in_redis(state: &SessionState, user_id: &str) -
...
+    redis_conn
+        .set_key_with_expiry(
+            key.as_str(),
+            user_attempts_value + 1,
+            consts::user::REDIS_TOTP_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+pub async fn check_totp_attempts_in_redis(state: &SessionState, user_id: &str) -> UserResult<u8> {
```
Comment (ThisIsMani) [inline crates/router/src/utils/user/two_factor_auth.rs:67]:
```suggestion

pub async fn get_totp_attempts_from_redis(state: &SessionState, user_id: &str) -> UserResult<u8> {
```

File: crates/router/src/routes/app.rs [hunk 5]
```diff
@@ -1850,6 +1850,10 @@ impl User {
         route = route.service(
             web::scope("/2fa")
                 .service(web::resource("").route(web::get().to(user::check_two_factor_auth_status)))
```
Comment (ThisIsMani) [inline crates/router/src/routes/app.rs:1852]:
```suggestion
// TODO: To be deprecated
.service(web::resource("").route(web::get().to(user::check_two_factor_auth_status)))
```

File: crates/router/src/core/user.rs [hunk 6]
```diff
@@ -1937,6 +1957,40 @@ pub async fn check_two_factor_auth_status(
...
+                is_completed: tfa_utils::check_totp_in_redis(&state, &user_token.user_id).await?,
+                remaining_attempts: consts::user::TOTP_MAX_ATTEMPTS
+                    - tfa_utils::check_totp_attempts_in_redis(&state, &user_token.user_id).await?,
+            },
+            recovery_code: user_api::TwoFactorAuthAttempts {
+                is_completed: tfa_utils::check_recovery_code_in_redis(&state, &user_token.user_id)
+                    .await?,
+                remaining_attempts: consts::user::RECOVERY_CODE_MAX_ATTEMPTS
+                    - tfa_utils::check_recovery_code_attempts_in_redis(&state, &user_token.user_id)
+                        .await?,
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1988]: NIT: It's better to split this into 2 variables like `totp` and `recovery_code` and construct this.

File: crates/router/src/core/errors/user.rs (hunk 1 above)
Comment (ThisIsMani) [reply crates/router/src/core/errors/user.rs:79]: And I think one error should be sufficient in this case, you can change this to something like this:
```rust
    MaxAttemptsReached(String), // String is the resource can be TOTP or RC.
```

**Review by ThisIsMani — COMMENTED** [review]
Should we remove the attempt keys from redis once the totp or recovery code is successful?

**Comment by ThisIsMani** [comment]
> Each key has an expiration time, after which it will be automatically removed. If the user still has remaining attempts, they can proceed directly to the next verification step.

Yeah the user might logic but we have 2FA verification inside the dashboard also... Won't this cause issue there?

File: crates/router/src/core/errors/user.rs [hunk 7]
```diff
@@ -269,6 +274,7 @@ impl UserErrors {
             Self::InvalidTotp => "Invalid TOTP",
             Self::TotpRequired => "TOTP required",
             Self::InvalidRecoveryCode => "Invalid Recovery Code",
+            Self::MaxAttemptsReached(error_message) => error_message,
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:277]:
```suggestion
            Self::MaxAttemptsReached(resource) =>
    format!("Max attempts reached for {}", resource"),
```

File: crates/router/src/core/user.rs [hunk 8]
```diff
@@ -1673,6 +1673,15 @@ pub async fn verify_totp(
         return Err(UserErrors::TotpNotSetup.into());
     }
 
+    if tfa_utils::get_totp_attempts_in_redis(&state, &user_token.user_id).await?
+        == consts::user::TOTP_MAX_ATTEMPTS
+    {
+        return Err(UserErrors::MaxAttemptsReached(
+            "Maximum TOTP attempts per user reached".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1680]:
```suggestion
        return Err(UserErrors::MaxAttemptsReached(
            "TOTP".to_string(),
```

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 9]
```diff
@@ -50,6 +50,69 @@ pub async fn check_recovery_code_in_redis(state: &SessionState, user_id: &str) -
...
+    redis_conn
+        .set_key_with_expiry(
+            key.as_str(),
+            user_attempts_value + 1,
+            consts::user::REDIS_TOTP_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+pub async fn get_totp_attempts_in_redis(state: &SessionState, user_id: &str) -> UserResult<u8> {
```
Comment (ThisIsMani) [inline crates/router/src/utils/user/two_factor_auth.rs:67]:
```suggestion
pub async fn get_totp_attempts_from_redis(state: &SessionState, user_id: &str) -> UserResult<u8> {
```

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 10]
```diff
@@ -50,6 +50,69 @@ pub async fn check_recovery_code_in_redis(state: &SessionState, user_id: &str) -
...
+        .set_key_with_expiry(
+            key.as_str(),
+            user_attempts_value + 1,
+            consts::user::REDIS_RECOVERY_CODE_ATTEMPTS_TTL_IN_SEC,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)
+}
+
+pub async fn get_recovery_code_attempts_in_redis(
```
Comment (ThisIsMani) [inline crates/router/src/utils/user/two_factor_auth.rs:99]:
```suggestion
pub async fn get_recovery_code_attempts_from_redis(
```

File: crates/router/src/core/user.rs [hunk 11]
```diff
@@ -1843,15 +1855,29 @@ pub async fn verify_recovery_code(
         return Err(UserErrors::TwoFactorAuthNotSetup.into());
     }
 
+    if tfa_utils::get_recovery_code_attempts_in_redis(&state, &user_token.user_id).await?
+        == consts::user::RECOVERY_CODE_MAX_ATTEMPTS
+    {
+        return Err(UserErrors::MaxAttemptsReached(
+            "Maximum recovery code attempts per user reached".to_string(),
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1862]:
```suggestion
            "Recovery Code".to_string(),
```

**Comment by Riddhiagrawal001** [comment]
>Yeah the user might logic but we have 2FA verification inside the dashboard also... Won't this cause issue there? 

It wont create issue just the user needs to wait till the cool-down period is over

File: crates/router/src/core/errors/user.rs [hunk 12]
```diff
@@ -90,6 +90,10 @@ pub enum UserErrors {
     SSOFailed,
     #[error("profile_id missing in JWT")]
     JwtProfileIdMissing,
+    #[error("MaxTotpAttemptsReached")]
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:93]: Please match the string here with the message in `get_error_message` function, like we are doing in some of the above errors.

File: crates/router/src/core/errors/user.rs (hunk 2 above)
Comment (ThisIsMani) [reply crates/router/src/core/errors/user.rs:282]: This got reverted again.

**Review by ThisIsMani — APPROVED** [review]
Minor concerns

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 13]
```diff
@@ -202,3 +139,88 @@ pub async fn delete_recovery_code_from_redis(
...
+        .map(|_| ())
+}
+
+pub async fn delete_recovery_code_attempts_from_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .delete_key(&get_recovery_code_attempts_key(user_id))
```
Comment (ThisIsMani) [inline crates/router/src/utils/user/two_factor_auth.rs:222]: Can you check if this breaks if key is not present in redis.

Comment (Riddhiagrawal001) [reply crates/router/src/utils/user/two_factor_auth.rs:222]: Tried for skip , it didn't throw any error

**Review by ThisIsMani — APPROVED** [review]

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 14]
```diff
@@ -139,3 +139,88 @@ pub async fn delete_recovery_code_from_redis(
...
+pub async fn delete_totp_attempts_from_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+    redis_conn
+        .delete_key(&get_totp_attempts_key(user_id))
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .map(|_| ())
```
Comment (apoorvdixit88) [inline crates/router/src/utils/user/two_factor_auth.rs:213]: If deletion fails, should we log the error here, with proper error message?

File: crates/router/src/core/user.rs [hunk 15]
```diff
@@ -1686,6 +1686,12 @@ pub async fn verify_totp(
         return Err(UserErrors::TotpNotSetup.into());
     }
 
+    if tfa_utils::get_totp_attempts_from_redis(&state, &user_token.user_id).await?
+        == consts::user::TOTP_MAX_ATTEMPTS
```
[the remark this answers is not in the document]
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs:1690]: After 4 attempts, the API will start returning an error, and after the 5-minute cooldown period, all the keys will expire, ensuring that it never exceeds 4 attempts. But for safety we can add

File: crates/router/src/utils/user/two_factor_auth.rs (hunk 14 above)
Comment (Riddhiagrawal001) [reply crates/router/src/utils/user/two_factor_auth.rs:215]: Added the log

**Review by apoorvdixit88 — DISMISSED** [review]

**Review by ThisIsMani — DISMISSED** [review]

File: crates/router/src/core/user.rs [hunk 16]
```diff
@@ -1924,10 +1944,17 @@ pub async fn terminate_two_factor_auth(
...
     }
 
-    let current_flow = domain::CurrentFlow::new(user_token, domain::SPTFlow::TOTP.into())?;
+    let current_flow = domain::CurrentFlow::new(user_token.clone(), domain::SPTFlow::TOTP.into())?;
     let next_flow = current_flow.next(user_from_db, &state).await?;
     let token = next_flow.get_token(&state).await?;
 
+    let _ = tfa_utils::delete_totp_attempts_from_redis(&state, &user_token.user_id)
+        .await
+        .map_err(|error| logger::error!(?error));
```
Comment (lsampras) [inline crates/router/src/core/user.rs:1953]: Do we need to handle this error? 
```suggestion
        .map_err(|error| logger::error!(?error))?;
```

File: crates/router/src/core/user.rs [hunk 17]
```diff
@@ -1924,10 +1944,17 @@ pub async fn terminate_two_factor_auth(
...
+    let _ = tfa_utils::delete_recovery_code_attempts_from_redis(&state, &user_token.user_id)
+        .await
+        .map_err(|error| logger::error!(?error));
```
Comment (lsampras) [inline crates/router/src/core/user.rs:1956]:
```suggestion
        .map_err(|error| logger::error!(?error))?;
```
?

File: crates/router/src/core/user.rs [hunk 18]
```diff
@@ -1856,15 +1865,26 @@ pub async fn verify_recovery_code(
...
-    let matching_index = utils::user::password::get_index_for_correct_recovery_code(
+    let Some(matching_index) = utils::user::password::get_index_for_correct_recovery_code(
         &req.recovery_code,
         &recovery_codes,
     )?
-    .ok_or(UserErrors::InvalidRecoveryCode)?;
+    else {
+        let _ = tfa_utils::insert_recovery_code_attempts_in_redis(&state, &user_token.user_id)
+            .await
+            .map_err(|error| logger::error!(?error));
```
Comment (lsampras) [inline crates/router/src/core/user.rs:1885]:
```suggestion
            .map_err(|error| logger::error!(?error))?;
```

File: crates/router/src/core/user.rs [hunk 19]
```diff
@@ -1702,6 +1708,9 @@ pub async fn verify_totp(
         .change_context(UserErrors::InternalServerError)?
         != req.totp.expose()
     {
+        let _ = tfa_utils::insert_totp_attempts_in_redis(&state, &user_token.user_id)
+            .await
+            .map_err(|error| logger::error!(?error));
```
Comment (lsampras) [inline crates/router/src/core/user.rs:1713]:
```suggestion
            .map_err(|error| logger::error!(?error))?;
```

**Review by lsampras — APPROVED** [review]
Some doubts about the error handling, 
Can ignore if it's intended that way

**Comment by Riddhiagrawal001** [comment]
@lsampras 

> Some doubts about the error handling, Can ignore if it's intended that way

We can omit this error propagation if the operation fails, as this is an additional check for maximum TOTP attempts. Even without this check, the main functionality should work correctly.

File: crates/router/src/core/errors/user.rs [hunk 20]
```diff
@@ -90,6 +90,10 @@ pub enum UserErrors {
     SSOFailed,
     #[error("profile_id missing in JWT")]
     JwtProfileIdMissing,
+    #[error("Maximum attempts reached for TOTP")]
+    MaxTotpAttemptsReached,
+    #[error("Maximum attempts reached for Recovery Code")]
+    MaxRecoveryCodeAttemptsReached,
```
[the remark this answers is not in the document]
Comment (Riddhiagrawal001) [reply crates/router/src/core/errors/user.rs:96]: As discussed, creating a common type will give them the flexibility to pass any value, and the output will always be of the form: `Max attempts reached for ${}`

File: crates/router/src/utils/user/two_factor_auth.rs [hunk 21]
```diff
@@ -139,3 +140,90 @@ pub async fn delete_recovery_code_from_redis(
...
+        .map(|v| v.unwrap_or(0))
+}
+
+pub async fn insert_recovery_code_attempts_in_redis(
+    state: &SessionState,
+    user_id: &str,
+) -> UserResult<()> {
+    let redis_conn = super::get_redis_connection(state)?;
+
+    let user_attempts_value = get_recovery_code_attempts_from_redis(state, user_id).await?;
```
Comment (racnan) [inline crates/router/src/utils/user/two_factor_auth.rs:182]: take this value as an argument.

File: crates/router/src/core/user.rs [hunk 22]
```diff
@@ -1856,15 +1865,26 @@ pub async fn verify_recovery_code(
...
         .ok_or(UserErrors::InternalServerError)?;
 
-    let matching_index = utils::user::password::get_index_for_correct_recovery_code(
+    let Some(matching_index) = utils::user::password::get_index_for_correct_recovery_code(
         &req.recovery_code,
         &recovery_codes,
     )?
-    .ok_or(UserErrors::InvalidRecoveryCode)?;
+    else {
+        let _ = tfa_utils::insert_recovery_code_attempts_in_redis(&state, &user_token.user_id)
```
Comment (racnan) [inline crates/router/src/core/user.rs:1883]: `get_recovery_code_attempts_from_redis` is also being called from `insert_recovery_code_attempts_in_redis`.

**Review by lsampras — DISMISSED** [review]

File: crates/router/src/core/user.rs [hunk 23]
```diff
@@ -1856,15 +1870,31 @@ pub async fn verify_recovery_code(
         return Err(UserErrors::TwoFactorAuthNotSetup.into());
     }
 
+    let user_totp_attempts =
+        tfa_utils::get_recovery_code_attempts_from_redis(&state, &user_token.user_id).await?;
+
+    if user_totp_attempts >= consts::user::RECOVERY_CODE_MAX_ATTEMPTS {
```
Comment (racnan) [inline crates/router/src/core/user.rs:1876]: incorrect variable name

File: crates/router/src/core/user.rs [hunk 24]
```diff
@@ -1686,6 +1686,13 @@ pub async fn verify_totp(
         return Err(UserErrors::TotpNotSetup.into());
     }
 
+    let user_recovery_code_attempts =
```
Comment (racnan) [inline crates/router/src/core/user.rs:1689]: incorrect variable name

**Review by racnan — APPROVED** [review]

**Review by racnan — DISMISSED** [review]
delete function calls are logging error at two places.

**Comment by Riddhiagrawal001** [comment]
@racnan @apoorvdixit88 
> delete function calls are logging error at two places.
Will remove the log from delete_recovery_code_attempts_from_redis &  delete_totp_attempts_from_redis function as error is logged where the function is called

File: crates/router/src/core/user.rs (hunk 23 above)
Comment (Riddhiagrawal001) [reply crates/router/src/core/user.rs]: Changed

**Review by ThisIsMani — APPROVED** [review]

**Review by racnan — APPROVED** [review]

**Review by apoorvdixit88 — APPROVED** [review]

**Pull request merged**